storage: Do not create nil data pointers in rocksdb slices

miretskiy opened the pull request:
Fixes #45524

Libroach's DBSlice data type gets converted to the rocksdb::Slice
whenever we need to access underlying rocksdb (iterators, get/set, etc).

However, internally, rocksdb::Slice asserts that the underlying data
pointer may not be null in the rocksdb::Slice::compare as observed in #45524

This change modifies our rocksdb bridge code to never generate
DBSlices with nullptr data.

Release notes: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45759)
<!-- Reviewable:end -->

itsbilal reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice catch!

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45759#-:-M1fvaYc-Mxew31uiWFf:b-pe67a)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @petermattis and @sumeerbhola)



<!-- Sent from Reviewable.io -->

itsbilal commented:
Spoke too soon. Looks like this is going to require some test changes as well as more len(result) != 0 checks in MVCC code. Happy to help with that.

miretskiy commented:
I see.  I'm going to try to make the change in libroach.h instead.

itsbilal commented:
Looking at the assertion, I think the most straightforward way might be to just remove it, instead of changing our code around for it. It's the only assertion I could find in rocksdb that has a check like this, and in all other cases, we look at `len` before diving into the data pointer. Even the rest of the `Slice::compare` code is `len`aware

We already extensively pass `nil` slices around between rocksdb and go, so I'm confident in being able to continue doing that.

miretskiy commented:
@petermattis  can you comment on this please?

So, the initial approach w/ changing rocksdb.go produced many test failures.

We have 3 alternatives:
 1.  Stick w/ the current approach & fix test failures.

 2. A change to libroach.h to modify 2 inline functions (I think this might be my preferred approach)

 inline rocksdb::Slice ToSlice(DBSlice s) { return rocksdb::Slice(s.data == nullptr ? "" : s.data, s.len); }
inline rocksdb::Slice ToSlice(DBString s) { return rocksdb::Slice(s.data == nullptr ? "" : s.data, s.len); }

  I suspect (given the fact that these are inline and copmare against nullptr) that the impact will be 0 (it should all get compiled away).

3. Remove assertion in our fork of rocksdb code.  That assertion is silly.  compare() works just fine
  w/ nullptr data.  We would have seen many failures if this weren't the case.
  Even with this assertion removed, I'm worried that there might be some implicit assumptions lurking
  in rocks code (e.g. that *slice.data is defined).

(And my least favorite, but I suppose a possibility as well, is to define a static emtpySlice sentinel
in libroach.h (empty DBSlice w/ data initialized to "").  This is not super desirable since, unfortunately, 
DBSlice is, in fact, not a constant data structure, and so there might be a chance that some
code might mutate data ptr).

petermattis reviewed (commented):
> 2. A change to libroach.h to modify 2 inline functions (I think this might be my preferred approach)

I would do this.

> (And my least favorite, but I suppose a possibility as well, is to define a static emtpySlice sentinel in libroach.h (empty DBSlice w/ data initialized to ""). This is not super desirable since, unfortunately, DBSlice is, in fact, not a constant data structure, and so there might be a chance that some code might mutate data ptr).

This sounds ok too, but sounds equivalent to option 2. Maybe I'm misunderstanding the problem you see. You'd define `static const rocksdb::Slice emptySlice("", 0)`. There would be no way to modify `emptySlice`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45759#-:-M1gO0siA6uQqcXcq2YF:b-t14o5j)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy and @sumeerbhola)

___
*[pkg/storage/rocksdb.go, line 2643 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45759#-M1gOh00CRG2wcU17cN-:-M1gOh00CRG2wcU17cN0:bpg1ukx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4a29bd3504368b9def97bf7fef3f05f9244a678/pkg/storage/rocksdb.go#L2643)):*
> ```Go
> // assertions checking for slice.data != nullptr.
> // Therefore, we represent an empty slice with an empty []byte array.
> var emptySlice = []byte{0}
> ```

I realize this approach doesn't work, but wanted to point out this is a bit of overkill. You're create a slice of length 1 in order to get a non-nil `*byte` pointer. There is an easier way:

```
var emptyByte byte
...
  if len(b) == 0 {
    return C.DBSlice(data: (*C.char)(&emptyByte), len : 0}
  }
```

This is equivalent to what you wrote, but just goes directly to a `*byte` rather than using a `[]byte`. The insight here is that `[]byte` is really just sugar for:

```
type byteSliceHeader {
  ptr *byte
  len int
  cap int
}
```


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
It's just I saw that, even though DBSlice should be an immutable data structure, the fact that we do use it
in a mutable way (e.g. chunked_buffer) made me a bit worried.  If you're okay w/ inline function changes, I'd do that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45759#-:-M1gQbmpEFskBscSfcn8:b7ktxdt)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @petermattis and @sumeerbhola)

___
*[pkg/storage/rocksdb.go, line 2643 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45759#-M1gOh00CRG2wcU17cN-:-M1gQpmVDVpZWXdjZuim:bw9rv7f) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4a29bd3504368b9def97bf7fef3f05f9244a678/pkg/storage/rocksdb.go#L2643)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I realize this approach doesn't work, but wanted to point out this is a bit of overkill. You're create a slice of length 1 in order to get a non-nil `*byte` pointer. There is an easier way:

```
var emptyByte byte
...
  if len(b) == 0 {
    return C.DBSlice(data: (*C.char)(&emptyByte), len : 0}
  }
```

This is equivalent to what you wrote, but just goes directly to a `*byte` rather than using a `[]byte`. The insight here is that `[]byte` is really just sugar for:

```
type byteSliceHeader {
  ptr *byte
  len int
  cap int
}
```
</blockquote></details>

Ahh.. Yeah... I didnt' try ptr *byte.  I tried creating e.g. []byte{} and pass in the value of that, but cgo was complaining.
Thanks for tip.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
The `libroach/db.h` change looks good, but don't you need to revert the `rocksdb.go` change? Or did you revert it and reviewable is messing up?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45759#-:-M1gXRuR3ZKP1EU5q34V:b-e5dfmy)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @petermattis and @sumeerbhola)



<!-- Sent from Reviewable.io -->

Commit:
storage: Do not create nil data pointers in rocksdb slices

Libroach's DBSlice data type gets converted to the rocksdb::Slice
whenever we need to access underlying rocksdb (iterators, get/set, etc).

However, internally, rocksdb::Slice asserts that the underlying data
pointer may not be null in the rocksdb::Slice::compare as observed in #45524

This change modifies our rocksdb bridge code to never generate
DBSlices with nullptr data.

Release notes: None

itsbilal reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> good work!

> Reviewed 1 of 1 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45759#-:-M1gYzR3-JYDIzrOHWpV:bc789to)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @petermattis and @sumeerbhola)



<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45759)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)



<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
> So, the initial approach w/ changing rocksdb.go produced many test failures.

I did not follow what was causing those test failures -- could you elaborate?

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45759#-:-M1hFxUv2SlA-QQpDhrX:baqovqk)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @sumeerbhola)



<!-- Sent from Reviewable.io -->

miretskiy commented:
Lots of assumptions that start key ("") is nil

On Thu, Mar 5, 2020 at 7:35 PM sumeerbhola <notifications@github.com> wrote:

> *@sumeerbhola* approved this pull request.
>
> So, the initial approach w/ changing rocksdb.go produced many test
> failures.
>
> I did not follow what was causing those test failures -- could you
> elaborate?
>
> [image: :lgtm:]
> <https://camo.githubusercontent.com/41b3ec3419116e3b3f84507ad965646ce4ce1f0c/68747470733a2f2f72657669657761626c652e696f2f6c67746d2e706e67>
>
> *Reviewable
> <https://reviewable.io/reviews/cockroachdb/cockroach/45759#-:-M1hFxUv2SlA-QQpDhrX:baqovqk>*
> status: [image: :shipit:] complete! 2 of 0 LGTMs obtained (waiting on
> @sumeerbhola <https://github.com/sumeerbhola>)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/45759?email_source=notifications&email_token=ANA4FVATRG6BBISP5TBGZH3RGBAMHA5CNFSM4LCPURF2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCYG6IPY#pullrequestreview-370009151>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANA4FVAFWT5D2P2NU6FQNRTRGBAMHANCNFSM4LCPURFQ>
> .
>

miretskiy commented:
bors r+

miretskiy commented:
Fixes #45524

knz commented:
bors r-
